### PR TITLE
Keccak PoC

### PR DESCRIPTION
**Pull request by querolita**
This PR contains the Keccak gadget for Snarky. It addresses the Keccak PoC. 

Closes:
- https://github.com/MinaProtocol/mina/issues/13046
- https://github.com/MinaProtocol/mina/issues/13160
- https://github.com/MinaProtocol/mina/issues/13161
- https://github.com/MinaProtocol/mina/issues/13188
- https://github.com/MinaProtocol/mina/issues/13191
- https://github.com/MinaProtocol/mina/issues/13190
- https://github.com/MinaProtocol/mina/issues/13163
- https://github.com/MinaProtocol/mina/issues/13168
- https://github.com/MinaProtocol/mina/issues/13167
- https://github.com/MinaProtocol/mina/issues/13166
- https://github.com/MinaProtocol/mina/issues/13165
- https://github.com/MinaProtocol/mina/issues/13164
- https://github.com/MinaProtocol/mina/issues/13169
- https://github.com/MinaProtocol/mina/issues/13158
- https://github.com/MinaProtocol/mina/issues/13159
- https://github.com/MinaProtocol/mina/issues/13162

Pending:
- https://github.com/MinaProtocol/mina/issues/13172

### CODE REVIEW

**Comment by querolita** [comment]
!ci-build-me

**Comment by querolita** [comment]
!ci-build-me

**Comment by querolita** [comment]
!ci-build-me

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+   * The padded message vector will start with the message vector
+   * followed by the 0*1 rule to fulfil a length that is a multiple of rate (in bytes)
+*)
+let pad_nist (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) (rate : int) : Circuit.Field.t list =
+  let open Circuit in
+  (* Find out desired length of the padding in bytes *)
+  (* If message is already rate bits, need to pad full rate again *)
+  let extra_bytes = (rate / 8) - (List.length message mod rate) in
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:244]: This can be negative if e.g. `List.length messages = (8*k+1)*rate+1`

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+  (* 0x06 0x00 ... 0x00 0x80 or 0x86 *)
+  let last_field = Common.two_pow (module Circuit) 7 in
+  let last = Field.constant @@ Common.two_pow (module Circuit) 7 in
+  (* Create the padding vector *)
+  let pad = Array.create ~len:extra_bytes Field.zero in
+  pad.(0) <- Field.of_int 6 ;
+  pad.(extra_bytes - 1) <- Field.add pad.(extra_bytes - 1) last ;
+  Field.Assert.equal (Field.constant last_field) last ;
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:252]: No-op, delete.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+  let open Circuit in
+  (* Find out desired length of the padding in bytes *)
+  (* If message is already rate bits, need to pad full rate again *)
+  let extra_bytes = (rate / 8) - (List.length message mod rate) in
+  (* 0x06 0x00 ... 0x00 0x80 or 0x86 *)
+  let last_field = Common.two_pow (module Circuit) 7 in
+  let last = Field.constant @@ Common.two_pow (module Circuit) 7 in
+  (* Create the padding vector *)
+  let pad = Array.create ~len:extra_bytes Field.zero in
+  pad.(0) <- Field.of_int 6 ;
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:250]: If `extra_bytes = 0` then this will fail.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+  (* 0x01 0x00 ... 0x00 0x80 or 0x81 *)
+  let last_field = Common.two_pow (module Circuit) 7 in
+  let last = Field.constant @@ last_field in
+  (* Create the padding vector *)
+  let pad = Array.create ~len:extra_bytes Field.zero in
+  pad.(0) <- Field.one ;
+  pad.(extra_bytes - 1) <- Field.add pad.(extra_bytes - 1) last ;
+  Field.Assert.equal (Field.constant last_field) last ;
+  (* Cast the padding array to a list *)
+  let pad = Array.to_list pad in
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:280]: Can we standardise on either `Array` or `List` in this module instead of mixing? Preferably `Array`

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+    (state : Circuit.Field.t State.matrix) : Circuit.Field.t State.matrix =
+  let open Circuit in
+  let state_a = state in
+  let state_c = Array.create ~len:5 Field.zero in
+  let state_d = Array.create ~len:5 Field.zero in
+  let state_e = State.zeros (module Circuit) in
+  (* XOR the elements of each row together *)
+  (* for all x in {0..4}: C[x] = A[x,0] xor A[x,1] xor A[x,2] xor A[x,3] xor A[x,4] *)
+  for x = 0 to keccak_dim - 1 do
+    state_c.(x) <-
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:303]: It looks like this could just be
```ocaml
let state_c = Array.map state_a ~f:(Array.reduce_exn ~f:(bxor64 (module Circuit)) in
```

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+             (bxor64
+                (module Circuit)
+                (bxor64 (module Circuit) state_a.(x).(0) state_a.(x).(1))
+                state_a.(x).(2) )
+             state_a.(x).(3) )
+          state_a.(x).(4))
+  done ;
+  (* for all x in {0..4}: D[x] = C[x-1] xor ROT(C[x+1], 1) *)
+  for x = 0 to keccak_dim - 1 do
+    state_d.(x) <-
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:318]: Prefer `Array.init`

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+      Bitwise.(
+        bxor64
+          (module Circuit)
+          state_c.(modulo (x - 1) keccak_dim)
+          (rot64 (module Circuit) state_c.((x + 1) mod keccak_dim) 1 Left)) ;
+    (* for all x in {0..4} and y in {0..4}: E[x,y] = A[x,y] xor D[x] *)
+    for y = 0 to keccak_dim - 1 do
+      state_e.(x).(y) <-
+        Bitwise.(bxor64 (module Circuit) state_a.(x).(y) state_d.(x))
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:327]:
```suggestion
let state_e =
  Array.map2_exn state_a state_d ~f:(fun state_a state_d ->
    Array.map state_a ~f:(Bitwise.bxor64 (module Circuit) state_d))
```

File: src/lib/crypto/kimchi_backend/gadgets/common.ml [hunk 8]
```diff
@@ -1,9 +1,9 @@
+
 (* Common gadget helpers *)
 
-open Core_kernel
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/common.ml:3]: It used to be there, but I removed it in order to be able to use the standard `List.of_seq` inside `field_bytes_of_hex` for which an equivalent does not seem to exist inside Core_kernel (or at least I couldn't find it).

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+(* Auxiliary function to check composition of 8 bytes into a 64-bit word *)
+let check_bytes_to_word (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (word : Circuit.Field.t) (word_bytes : Circuit.Field.t list) =
+  let open Circuit in
+  let composition =
+    List.foldi word_bytes ~init:Field.zero ~f:(fun i acc x ->
+        let shift = Field.constant @@ Common.two_pow (module Circuit) (8 * i) in
+        (* TODO: necessary step to check values? *)
+        Field.Assert.equal
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:80]: I was not sure if Snarky was already constraining it (that I am using this shift to compose 8 bytes into 64-bit words), but if I delete that Field.Assert.equal, the same circuit results. So it seems like it is indeed redundant/unnecessary and thus no-op. Deleting now.

[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:80]: thus the comment `(* TODO: necessary step to check values? *)` , so answer is no, not necessary 👍🏻

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 10]
```diff
@@ -0,0 +1,720 @@
...
+        for z = 0 to (keccak_word / 8) - 1 do
+          let index = (8 * ((keccak_dim * y) + x)) + z in
+          (* Field element containing value 2^(8*z) *)
+          let shift_field =
+            Common.bignum_bigint_to_field
+              (module Circuit)
+              Bignum_bigint.(pow (of_int 2) (of_int (Int.( * ) 8 z)))
+          in
+          let shift = Field.constant shift_field in
+          (* TODO: Does this generate automatic generic gates to check composition? *)
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:132]: Deleting here as well, it does nothing

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 11]
```diff
@@ -0,0 +1,720 @@
...
+    state
+
+  (* Converts a state of cvars to a list of bytes as cvars *)
+  let as_prover_to_bytes (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (state : Circuit.Field.t matrix) : Circuit.Field.t list =
+    let open Circuit in
+    assert (
+      Array.length state = keccak_dim && Array.length state.(0) = keccak_dim ) ;
+    let bytestring = Array.create ~len:200 Field.zero in
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:148]: oh okay, comes from 1600 = keccak state / 8 bits per byte. I will create a constant for it

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 12]
```diff
@@ -0,0 +1,721 @@
...
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:148]: Oh okay thank you let me try! I am indeed overusing imperative style in here 😅

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 13]
```diff
@@ -0,0 +1,720 @@
...
+      done
+    done ;
+    Array.to_list bytestring
+
+  let xor (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (input1 : Circuit.Field.t matrix) (input2 : Circuit.Field.t matrix) :
+      Circuit.Field.t matrix =
+    let open Circuit in
+    assert (
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:195]: Quite elegant, nice! Imperativeness really is the first idea in my mind

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+      Array.length input1 = keccak_dim && Array.length input1.(0) = keccak_dim ) ;
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:196]: Same as above, I just had them so that a future circuit writer does not misuse it. Those are just normal asserts, they don't add rows to the circuit. I believed it was just a cheap sanity check. But I can remove it and instead add a disclaimer in front of the function to explain how it should be called.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+    in
+    for y = 0 to keccak_dim - 1 do
+      for x = 0 to keccak_dim - 1 do
+        output.(x).(y) <-
+          Bitwise.bxor64 (module Circuit) input1.(x).(y) input2.(x).(y)
+      done
+    done ;
+    output
+
+  let _print (type f)
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:210]: Oh, yeah true, I was using a version of it using some prints inside the loop and removed it during debugging for the committed version. Right now it is not even being used (`_`). I will just delete this function and only recover it again from the history of commits if needed.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 16]
```diff
@@ -0,0 +1,721 @@
...
+            in
+            ()
+          done
+        done )
+end
+
+(* KECCAK HASH FUNCTION IMPLEMENTATION *)
+
+(* Performs the modulo operation, not remainder as in mod which instead computes integer remainder *)
+let modulo (x : int) (m : int) : int = ((x mod m) + m) mod m
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:230]: Oh well, true. I can remove it and use that instead.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml (hunk 1 above)
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:244]: Okay the longest test I had had 455 bits. Now I have one with more than 1088 bits to force it not only use more blocks but also break the typo-ed formula when padding. The new test uses 30k rows instead (2 blocks not only 1). Fixed now, both in nist and 10*1 (they were a copy paste of each other).

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 17]
```diff
@@ -0,0 +1,686 @@
...
+                   ; w5 = List.nth_exn lookups ((3 * i) + 2)
+                   ; w6 = Field.zero
+                   } )
+          } ) ;
+    ()
+  done ;
+  ()
+
+(*
+* Keccak hash function with input message passed as list of 1byte Cvars.
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:539]: "Cvar bytes"

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 18]
```diff
@@ -0,0 +1,686 @@
...
+      let cs, _proof_keypair, _proof =
+        Runner.generate_and_verify_proof ?cs (fun () ->
+            let open Runner.Impl in
+            assert (String.length message % 2 = 0) ;
+            let message =
+              Array.to_list
+              @@ exists
+                   (Typ.array ~length:(String.length message / 2) Field.typ)
+                   ~compute:(fun () ->
+                     Array.of_list
```
Comment (jspada) [inline src/lib/crypto/kimchi_backend/gadgets/keccak.ml:613]: It would be good to line it up so everything is either `List` or `Array` in order avoid overhead of swapping back and forth.

**Review by jspada — CHANGES_REQUESTED** [review]
What a milestone!  Keccak in zero-knowledge :-)  Amazing work

Noticed a potential soundness issue and one potential bug.

Left a bunch of other suggestions.  I think you should add some constrain system reuse tests; some negative tests and edge-case tests.

File: src/lib/crypto/kimchi_backend/gadgets/common.ml [hunk 19]
```diff
@@ -185,8 +184,41 @@ let field_of_hex (type f)
     (hex : string) : f =
   bignum_bigint_to_field (module Circuit) @@ bignum_bigint_of_hex hex
 
+(* List of field elements for each byte of hexadecimal input*)
+let field_bytes_of_hex (type f)
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/common.ml:190]: It should complain because it needs a well formed hex, since we are not giving support for partial bits only full bytes. I will add tests capturing this

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 20]
```diff
@@ -0,0 +1,686 @@
...
+* Keccak hash function with input message passed as list of 1byte Cvars.
+* The message will be parsed as follows:
+* - the first byte of the message will be the least significant byte of the first word of the state (A[0][0])
+* - the 10*1 pad will take place after the message, until reaching the bit length rate.
+* - then, {0} pad will take place to finish the 1600 bits of the state.
+*)
+let hash (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) ~(length : int) ~(capacity : int)
+    (nist : bool) : Circuit.Field.t list =
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:548]: I will go for `nist_version`

File: src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml [hunk 21]
```diff
@@ -1127,6 +1127,51 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
...
+   * This is partly implicit in the circuit given the number of byte outputs of Keccak:
+   * · input_bitlen < fmod_bitlen : OK
+   * · input_bitlen = fmod_bitlen : OK
+   * · input_bitlen > fmod_bitlen : CONSTRAIN
+   * Check that the most significant byte of the input is less than 2^(fmod_bitlen % 8)
+   *)
+  let input_bitlen = Array.length bytestring * 8 in
+  if input_bitlen > fmod_bitlen then
+    (* For the most significant one, constrain that it is less bits than required *)
+    Lookup.less_than_bits (module Circuit) (fmod_bitlen % 8) bytestring.(0) ;
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml:1147]: I mean, I guess this is part of the integration tests we want to do. It could be anything, I don't know what's the standard in other ECDSA implementations. We should compare this when we know more, in any case good to leave this open.

File: src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml [hunk 22]
```diff
@@ -1127,6 +1127,51 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
...
+  (* C2: Constrain bytes into standard foreign field element limbs => foreign field element z *)
+  let limbs =
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml:1149]: Well, not really. That gives me the element as an `Element.Standard` and I want access to each of the limbs separately

[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml:1149]: But then I need an `exists` scope and I can only return Field.typ or arrays from there.

File: src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml [hunk 23]
```diff
@@ -1127,6 +1127,51 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
...
+   *   Constrain z' = z + 0 modulo foreign_field_modulus using foreign field addition gate
+   *
+   *   Note: this is sufficient because z cannot be double the size due to bit length constraint
+   *)
+  (* C4: Range check z' < f *)
+  (* Altogether this is a call to Foreign_field.add in default mode *)
+  let elem = Element.Standard.of_limbs (limbs.(0), limbs.(1), limbs.(2)) in
+  let zero = Element.Standard.of_limbs (Field.zero, Field.zero, Field.zero) in
+
+  let output = add (module Circuit) elem zero fmod in
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml:1170]: Not really, because the `add` gadget in default mode already takes care of all of that. From the docs:
```
 * In default mode:
 *     It adds a FFAdd gate,
 *     followed by a Zero gate,
 *     a FFAdd gate for the bound check,
 *     a Zero gate after this bound check,
 *     and a Multi Range Check gadget.
 ```

File: src/lib/crypto/kimchi_backend/gadgets/lookup.ml [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+     *   Inputs value to be checked and number of bits
+     *   Returns true if constraints are satisfied, false otherwise.
+     *)
+    let test_lookup ?cs ~bits value =
+      (* Generate and verify proof *)
+      let cs, _proof_keypair, _proof =
+        Runner.generate_and_verify_proof ?cs (fun () ->
+            let open Runner.Impl in
+            (* Set up snarky variable *)
+            let value =
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/lookup.ml:69]: but if it is a constant then you don't need to constrain anything, no? what are you suggesting exactly?

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 25]
```diff
@@ -0,0 +1,720 @@
...
+let keccak_ell = 6
+
+(* width of the lane of the state, meaning the length of each word in bits (64) *)
+let keccak_word = Int.pow 2 keccak_ell
+
+(* length of the state in bits, meaning the 5x5 matrix of words in bits (1600) *)
+let keccak_state_length = Int.pow keccak_dim 2 * keccak_word
+
+(* number of rounds of the Keccak permutation function depending on the value `l` (24) *)
+(*let keccak_rounds = 12 + (2 * keccak_ell)*)
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:24]: Oh sure, that was commented because at first I was manually changing the number of rounds to be 1 or 2, but now there is no need to have it like that anymore.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 26]
```diff
@@ -0,0 +1,686 @@
...
+    let state_length_in_bytes = keccak_state_length / 8 in
+    let bytes_per_word = keccak_word / 8 in
+    let bytestring =
+      Array.init state_length_in_bytes ~f:(fun idx ->
+          exists Field.typ ~compute:(fun () ->
+              (* idx = z + 8 * ((dim * y) + x) *)
+              let z = idx % bytes_per_word in
+              let x = idx / bytes_per_word % keccak_dim in
+              let y = idx / bytes_per_word / keccak_dim in
+              (*  [7 6 5 4 3 2 1 0] [x=0,y=1] [x=0,y=2] [x=0,y=3] [x=0,y=4]
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:151]: No, I was just trying to illustrate the order of words in the 5x5 matrix, and then inside each of them, the order of each byte (the [7 6 ... 0] is the 0,0 cell

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 27]
```diff
@@ -0,0 +1,686 @@
...
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f) :
+      Circuit.Field.t matrix =
+    let open Circuit in
+    let state =
+      Array.make_matrix ~dimx:keccak_dim ~dimy:keccak_dim Field.zero
+    in
+    state
+
+  (* Updates the cells of a state with new values *)
+  let update (type f)
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:99]: Oh didn't know that okay

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 28]
```diff
@@ -0,0 +1,686 @@
...
+   * The padded message vector will start with the message vector
+   * followed by the 0*1 rule to fulfil a length that is a multiple of rate (in bytes)
+*)
+let pad_nist (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) (rate : int) : Circuit.Field.t list =
+  let open Circuit in
+  (* Find out desired length of the padding in bytes *)
+  (* If message is already rate bits, need to pad full rate again *)
+  let extra_bytes = (rate / 8) - (List.length message mod (rate / 8)) in
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:211]: Rate cannot be < 8 because it has fixed values: 1600 - capacity and capacity is positive and smaller than 1600

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 29]
```diff
@@ -0,0 +1,686 @@
...
+    let state_i = round (module Circuit) state i in
+    (* Update state for next step *)
+    State.update (module Circuit) ~prev:state ~next:state_i
+  done ;
+  state
+
+(* Absorb padded message into a keccak state with given rate and capacity *)
+let absorb (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (padded_message : Circuit.Field.t list) ~(capacity : int) ~(rate : int) :
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:409]: Yeah, that is done later in the `hash` function (later in the code, before in the execution), because absorb is called from hash

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 30]
```diff
@@ -0,0 +1,686 @@
...
+  let hashed = Array.sub output_array ~pos:0 ~len:(length / 8) in
+
+  Array.to_list hashed
+
+(* Keccak sponge function for 1600 bits of state width
+ * Need to split the message into blocks of 1088 bits. 
+ *)
+let sponge (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (padded_message : Circuit.Field.t list) ~(length : int) ~(capacity : int)
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:484]: same, done in hash

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 31]
```diff
@@ -0,0 +1,686 @@
...
+    Core_kernel.List.map ~f:(fun x -> Field.(of_int 16 * x)) inputs
+  in
+  (* We need to lookup that both the inputs and the shifted values are less than 12 bits *)
+  (* Altogether means that it was less than 8 bits *)
+  let lookups = inputs @ shifted in
+  (* Make sure that a multiple of 3 cvars is in the list *)
+  let lookups =
+    match List.length lookups % 3 with
+    | 2 ->
+        lookups @ [ Field.zero; Field.zero ]
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:511]: Oh sure, good catch

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 32]
```diff
@@ -0,0 +1,686 @@
...
+    | _ ->
+        lookups
+  in
+  (* We can fit 3 12-bit lookups per row *)
+  for i = 0 to (List.length lookups / 3) - 1 do
+    with_label "lookup_byte" (fun () ->
+        assert_
+          { annotation = Some __LOC__
+          ; basic =
+              Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:523]: (also, the gadget came after this code was done)

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 33]
```diff
@@ -0,0 +1,686 @@
...
+          let index = (8 * ((keccak_dim * y) + x)) + z in
+          (* Field element containing value 2^(8*z) *)
+          let shift_field =
+            Common.bignum_bigint_to_field
+              (module Circuit)
+              Bignum_bigint.(pow (of_int 2) (of_int (Int.( * ) 8 z)))
+          in
+          let shift = Field.constant shift_field in
+          state.(x).(y) <-
+            Field.(state.(x).(y) + (shift * List.nth_exn bytestring index))
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:129]: Or use sub arrays with better complexity I guess

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 34]
```diff
@@ -0,0 +1,686 @@
...
+  let open Circuit in
+  (* Round constants for this round for the iota algorithm *)
+  let rc =
+    exists Field.typ ~compute:(fun () ->
+        Common.field_of_hex (module Circuit) round_consts.(round) )
+  in
+  let state_g = state in
+  state_g.(0).(0) <- Bitwise.(bxor64 (module Circuit) state_g.(0).(0) rc) ;
+  (* Check it is the right round constant *)
+  Field.Assert.equal rc
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:374]: I mean you only do this once per round, since there are 24 constants. If you want I can try to make it even before in the circuit, so that multiple blocks don't create more of these. But "again and again each time" is not many times fortunately.

[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:374]: Yeah I am checking now. In 2 blocks, doing so means saving 10 rows. Not amazing, but good to have. Thanks for the idea, I will push it like that.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml (hunk 18 above)
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:613]: Yeah but each of them have their own limitations. For example, with array it is cheap to access a certain position and modify its value. But for list, it is cheap to append values, but expensive to access a certain position. And Keccak has both things: you need to access values to compose words, but need the ability to put together when padding. It feels like no matter what path you take, it will have overheads in any case. After reading the code, what do you think makes better sense? Array always?

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 35]
```diff
@@ -0,0 +1,686 @@
...
+  let bytes_per_squeeze = rate / 8 in
+  (* number of squeezes *)
+  let squeezes = (length / rate) + 1 in
+  (* multiple of rate that is larger than output_length, in bytes *)
+  let output_length = squeezes * bytes_per_squeeze in
+  (* array with sufficient space to store the output *)
+  let output_array = Array.create ~len:output_length Field.zero in
+  (* first state to be squeezed *)
+  let bytestring = State.as_prover_to_bytes (module Circuit) state in
+  let output_bytes = List.take bytestring bytes_per_squeeze in
```
[the remark this answers is not in the document]
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:460]: Well, bytestring will always be 200 bytes in length. And bytes per squeeze is rate/8, and rate is  always smaller than 200. Thus the above question is yes

**Comment by querolita** [comment]
!ci-build-me

**Comment by querolita** [comment]
!ci-build-me

File: src/lib/crypto/kimchi_backend/gadgets/lookup.ml (hunk 24 above)
[the remark this answers is not in the document]
Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/lookup.ml:73]: well constants get added as generic gates which can sometimes be problematic if snarky adds them between chained gates.  I'm just suggesting to run a test with it as a circuit constant to catch any potential issue like this

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml (hunk 18 above)
Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:603]: It sounds like you've done the best possible since you're faced with both these requirements and limitations.

File: src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml (hunk 23 above)
[the remark this answers is not in the document]
Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml:1173]: Ah!  Well that's good then!

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml [hunk 36]
```diff
@@ -0,0 +1,686 @@
...
+  pad.(extra_bytes - 1) <- Field.add pad.(extra_bytes - 1) last ;
+  Field.Assert.equal (Field.constant last_field) last ;
+  (* Cast the padding array to a list *)
+  let pad = Array.to_list pad in
+  (* Return the padded message *)
+  message @ pad
+
+(* Pads a message M as:
+   * M || pad[x](|M|)
+   * Padding rule 10*1.
```
[the remark this answers is not in the document]
Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:230]: ok - that's what I thought.  Maybe also put it in english.

File: src/lib/crypto/kimchi_backend/gadgets/keccak.ml (hunk 3 above)
Comment (querolita) [reply src/lib/crypto/kimchi_backend/gadgets/keccak.ml:221]: No, it could be larger and that's why we use mod there. If it we have more input bytes than rate/8 bytes, then that means we need more blocks, and thus 15k more rows. In any case, the padding is meant to pad up to a multiple of rate bits.

**Comment by querolita** [comment]
!ci-build-me

**Review by jspada — APPROVED** [review]

**Comment by querolita** [comment]
!ci-build-me
